pyproject_metadata/__init__.py

```diff
@@ -543,13 +540,15 @@ def _write_metadata(  # noqa: C901
         """
         Write the metadata to the message. Handles JSON or Message.
         """
-        self.validate(warn=False)
+        errors = ErrorCollector(collect_errors=self.all_errors)
+        with errors.collect():
+            self.validate(warn=False)
 
         smart_message["Metadata-Version"] = self.auto_metadata_version
         smart_message["Name"] = self.name
         if not self.version:
-            msg = "Missing version field"
-            raise ConfigurationError(msg)
+            msg = "Field {key} missing"
+            errors.config_error(msg, key="project.version")
```

Comment:
Another inconsistency in error messages style. The errors messages below put the field name at the end after columns. This one puts it in the prose. For consistency with the errors below, this should be `'Field missing: {key!r}'`. But I much prefer the style used here, thus I would make the errors below match.